lib/pi_piper/pin.rb

```diff
@@ -105,8 +100,8 @@ def update_value(new_value)
     # @param [Symbol] state Indicates if and how pull mode must be set when 
     # pin direction is set to :in. Either :up, :down or :offing. Defaults to :off.
     def pull!(state)
-      return nil if @direction != :in
-      fail PiPiper::PinError, "Pin #{@pin} already released" if released?
+     raise PiPiper::PinError, "Unable to use pull-ups : pin direction must be ':in' for this" if 
+        @direction != :in and [:up, :down].include?(state)
```

Comment:
I didn't change a thing in this method. 
my opinion is that pull! & others, should only be called on initialize & initialize should validate every arguments it receives. we are far from this now, but i'm ok to do it in the rewrite.